### ISSUE TITLE
Compiler generate buggy code when using exception filter with pattern matching 

### ISSUE THREAD

**Issue by mkjeff** [dotnet/roslyn#59050]
**Version Used**: 
9.0+
**Steps to Reproduce**:
```cs
 static class C
 {
      static async Task Main()
      {
           Debug.Assert(await ExceptionFilterBroken());
      }

     public static async Task<bool> ExceptionFilterBroken()
     {
         try
         {
             await ThrowException();
             return true;
         }
         catch (Exception ex) when (ex.InnerException is ApplicationException { Message: "bad dog" or "dog bad" })
         {
            return await FooAsync();
         }
         catch
         {
            return false;
         }
     }

     static Task ThrowException() => throw new Exception("", new ApplicationException("bad dog"));
     static Task<bool> FooAsync() => Task.FromResult(true);
}
```
**Expected Behavior**:
   Debug.Assert without error

**Actual Behavior**:
   Debug.Assert failed
![image](https://user-images.githubusercontent.com/15941979/150926857-7c0f0943-383d-428f-bf69-1deef4302906.png)

I have tried certain cases to identify the issue in my [repo](https://github.com/mkjeff/CsExceptionFilterBug)

**Comment by jcouv** [comment]
Thanks @mkjeff. The compact repro and disassembled code were very helpful. 

I confirmed the issue and have a fix under review.
The problem occurs when there's an `await` in `catch` and the exception filter for the `catch` uses the exception and is complicated enough that it "spills" (we have a list [here](https://github.com/dotnet/roslyn/pull/46054#discussion_r456699320)).
In such situation, the exception filter sees `null` instead of the proper exception reference.

**Comment by jcouv** [comment]
Fix was merged for 17.2 (preview 2 I think)
